### PR TITLE
[SPARK-25425][SQL] Extra options should override session options in DataSource V2

### PR DESCRIPTION
**Pull request by MaxGekk**
## What changes were proposed in this pull request?

In the PR, I propose overriding session options by extra options in DataSource V2. Extra options are more specific and set via `.option()`, and should overwrite more generic session options. Entries from seconds map overwrites entries with the same key from the first map, for example:
```Scala
scala> Map("option" -> false) ++ Map("option" -> true)
res0: scala.collection.immutable.Map[String,Boolean] = Map(option -> true)
```

## How was this patch tested?

Added a test for checking which option is propagated to a data source in `load()`.

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
@MaxGekk . This looks worth for JIRA issue. Could you file a JIRA issue for this?

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala]: Also, cc @rdblue since this is introduced at https://github.com/apache/spark/commit/aadf9535b4a11b42fd9d72f636576d2da0766199#diff-f70bda59304588cc3abfa3a9840653f4R197 .

Comment (MaxGekk) [reply sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala]: @dongjoon-hyun The commit didn't change semantic actually. Before it was:
```
(extraOptions ++
        DataSourceV2Utils.extractSessionConfigs(
          ds = ds.asInstanceOf[DataSourceV2],
          conf = sparkSession.sessionState.conf))
```

Comment (dongjoon-hyun) [reply sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala]: Oh. It has more history. Thanks, @MaxGekk . Could you trace down when it started? We need to mark the affected version correctly in order to know the backport candidates.

Comment (dongjoon-hyun) [reply sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala]: Also, please add a test case for this. If this has a long history than SPARK-23203 (fixed at 2.4.0), we need to verify this during backporting.

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: Both read/write-side tests.

**Comment by maropu** [comment]
retest this please

**Review by HyukjinKwon — APPROVED** [review]
Looks good but let me leave it to @cloud-fan

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala
Comment (HyukjinKwon) [reply sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: +1 for tests

Comment (HyukjinKwon) [reply sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: +1 for tests

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala
Comment (MaxGekk) [reply sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala]: The changes were added in https://github.com/apache/spark/pull/19861

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala
Comment (MaxGekk) [reply sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: I wrote a test for read path since I was able to grab options propagated to DataSource but I have no idea so far for write path, only mocking probably. Does it make sense to do that?

**Comment by cloud-fan** [comment]
good catch! I believe this was a mistake. LGTM

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala
Comment (dongjoon-hyun) [reply sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala]: Thank you. Then, it was 2.3.0.

File: sql/core/src/test/scala/org/apache/spark/sql/sources/v2/DataSourceV2Suite.scala
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/v2/DataSourceV2Suite.scala]: `DataSourceWithSessionConfV2` -> `DataSourceV2WithSessionConfig`?

**Comment by dongjoon-hyun** [comment]
@MaxGekk . Could you specify `DSv2` more clearly in the PR title and description because it happens DSv2 code path only?

File: sql/core/src/test/scala/org/apache/spark/sql/sources/v2/DataSourceV2Suite.scala
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/v2/DataSourceV2Suite.scala]: Ur, it seems that we can use the existing one.

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala
Comment (dongjoon-hyun) [reply sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: For write path, I think we can use the traditional way instead of introducing mocking. Let me try.

**Comment by dongjoon-hyun** [comment]
@MaxGekk . I made a write-path test case PR to you. Could you review that, https://github.com/MaxGekk/spark-1/pull/8 ?

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM pending Jenkins tests.

**Comment by dongjoon-hyun** [comment]
Oh, it was @MaxGekk . I helped some test cases.
BTW, @MaxGekk . Could you send a backport PR to `branch-2.4`?

**Comment by MaxGekk** [comment]
> There is some conflict. Please make backporting PRs to 2.4 and 2.3.
> BTW, @MaxGekk . Could you send a backport PR to branch-2.4?

Sure, I will do that.

**Comment by dongjoon-hyun** [comment]
Thank you so much, @MaxGekk . It would be great if we can have that Spark 2.4.0 RC2.

**Comment by rdblue** [comment]
Thanks @MaxGekk, sorry for the original omission!

**Comment by MaxGekk** [comment]
@dongjoon-hyun Here is the PR for 2.4: https://github.com/apache/spark/pull/22474 . I will check does it have conflicts in 2.3 and if so, I will backport it to 2.3

**Comment by MaxGekk** [comment]
@dongjoon-hyun It cannot be merged to 2.3 easily because `DataSourceV2Relation` doesn't have the `options` field, and the test for read is not compilable. I will try to fix it tomorrow.

**Comment by dongjoon-hyun** [comment]
I see. Thank you, @MaxGekk !

**Pull request closed without merging**
